util/spearman.hpp

```diff
@@ -5,22 +5,20 @@
 #include <cmath>
 using namespace std;
 
-// Function returns the rank vector of a set of observations v
+// Function returns the 1-based rank vector of a set of observations v
 template <typename T>
-std::vector<double> rankify(std::vector<T> &v) {
+std::vector<double> rankify(const std::vector<T> &v) {
+    std::vector<T> sorted = v;
+    std::sort(begin(sorted), end(sorted));
+
     std::vector<double> result(v.size());
 
     for (size_t i = 0; i < v.size(); i++) {
-        size_t r = 1, s = 1;
-
-        for (size_t j = 0; j < v.size(); j++) {
-            if (v[j] < v[i])
-                r++;
-            if (v[j] == v[i])
-                s++;
-        }
+        const auto lb = std::lower_bound(std::begin(sorted), std::end(sorted), v[i]);
+        const auto ub = std::upper_bound(std::begin(sorted), std::end(sorted), v[i]);
+        const size_t r = 1 + (lb - std::begin(sorted)), s = ub - lb;
 
```

Comment:
This for loops counts `r`, the number of elements smaller than `v[i]`, and `s`, the number of elements equals to `v[i]`.
(Actually, I think `s=2` for unique elements since it's initialized to `1` already.)

In the new code, `lb` and `ub` point to the first element equal to `v[i]` and the first element larger than `v[i]` in the sorted list of values. Hence, `ub-lb` is the number of elements equal to `v[i]`, and `1 + lb - begin(sorted)` is the number of elements less than `v[i]` plus one.

Reply:
oh yeah, I thought the inner for loop goes until index `i`, in that case yeah I think your code is equivalent.